### PR TITLE
CR-3430: [CFN] public_access_security_groups_all_ports_rule

### PR DESCRIPTION
**Pull request by Imanuel-Mizrahi**

### CODE REVIEW

File: cloudrail/knowledge/utils/utils.py [hunk 1]
```diff
@@ -137,6 +137,8 @@ def flat_list(list_of_lists: List[list]) -> list:
 def hash_list(data: list) -> int:
     return hash(str(data))
 
+def check_array_has_value(array: Union[list, set]) -> bool:
+    return bool(array) and all(x for x in array)
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/utils.py:141]: what if the array has value of False ?

File: cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py [hunk 2]
```diff
@@ -17,8 +17,22 @@ def parse_resource(self, cfn_res_attr: dict) -> Ec2Instance:
...
         network_interfaces = self.get_property(properties, 'NetworkInterfaces', [])
-        network_interface_ids = [ni['NetworkInterfaceId'] for ni in network_interfaces] or self.CFN_PSEUDO_LIST
-        security_groups_ids = self.get_property(properties, 'SecurityGroupIds')
+        network_interface_ids = [ni.get('NetworkInterfaceId') for ni in network_interfaces]
+        network_interface_ids = network_interface_ids if check_array_has_value(network_interface_ids) else self.CFN_PSEUDO_LIST
+        security_group_ids_from_enis = flat_list([ni.get('GroupSet') for ni in network_interfaces])
+        security_group_ids_from_enis = set(sg for sg in security_group_ids_from_enis)
+        security_groups_ids_from_resource = self.get_property(properties, 'SecurityGroupIds')
+        security_groups_ids_from_resource = security_groups_ids_from_resource \
+            if check_array_has_value(security_groups_ids_from_resource) and all('.' not in x for x in security_groups_ids_from_resource) else None
```
Comment (Tomerva) [inline cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py:26]: what if not **all**, but some values have '.' ?

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/utils/utils.py [hunk 3]
```diff
@@ -137,6 +137,8 @@ def flat_list(list_of_lists: List[list]) -> list:
 def hash_list(data: list) -> int:
     return hash(str(data))
 
+def check_array_has_value(array: Union[list, set]) -> bool:
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/utils.py:140]: this type is too restrictive, and the name is not accurate
it should be of type Iterable, and the arg name and function name should change accordingly as well.

also please add tests for this function

(an iterable is any collection which you can **iterate** over, like run for-loops on, or for-comprehension)

File: cloudrail/knowledge/utils/utils.py [hunk 4]
```diff
@@ -137,6 +137,8 @@ def flat_list(list_of_lists: List[list]) -> list:
...
+    return bool(array) and all(x for x in array if not isinstance(x, bool))
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/utils.py:141]: if the function checks that a value exists, which im guessing for a list like [ None, None ] it should return false, i would implement it like this:

`return any[x is not None for x in iterable]`

File: cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py [hunk 5]
```diff
@@ -17,8 +17,22 @@ def parse_resource(self, cfn_res_attr: dict) -> Ec2Instance:
...
         network_interfaces = self.get_property(properties, 'NetworkInterfaces', [])
-        network_interface_ids = [ni['NetworkInterfaceId'] for ni in network_interfaces] or self.CFN_PSEUDO_LIST
-        security_groups_ids = self.get_property(properties, 'SecurityGroupIds')
+        network_interface_ids = [ni.get('NetworkInterfaceId') for ni in network_interfaces]
+        network_interface_ids = network_interface_ids if check_array_has_value(network_interface_ids) else self.CFN_PSEUDO_LIST
+        security_group_ids_from_enis = flat_list([ni.get('GroupSet') for ni in network_interfaces])
+        security_group_ids_from_enis = set(sg for sg in security_group_ids_from_enis)
+        security_groups_ids_from_resource = self.get_property(properties, 'SecurityGroupIds')
+        security_groups_ids_from_resource = security_groups_ids_from_resource \
+            if check_array_has_value(security_groups_ids_from_resource) and any('.' not in x for x in security_groups_ids_from_resource) else None
```
Comment (Tomerva) [inline cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py:26]: where you use check_array_has_value, wouldnt you want to get only the not-None values?

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py (hunk 5 above)
Comment (Tomerva) [reply cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py]: what about the other places?

File: tests/unit/utils/test_list_utils.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+import unittest
+
+from cloudrail.knowledge.utils.utils import check_iterable_has_value
+
+
+class TestListUtils(unittest.TestCase):
```
Comment (Tomerva) [inline tests/unit/utils/test_list_utils.py:6]: please rename class (and file), as its not a List util
you can rename it to like TestCheckIterableHasValue and this file will only test this function

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py [hunk 7]
```diff
@@ -17,8 +17,21 @@ def parse_resource(self, cfn_res_attr: dict) -> Ec2Instance:
         public_ip = self.get_property(properties, 'PublicIp')
         ipv6_addresses = self.get_property(properties, 'Ipv6Addresses', [])
         network_interfaces = self.get_property(properties, 'NetworkInterfaces', [])
-        network_interface_ids = [ni['NetworkInterfaceId'] for ni in network_interfaces] or self.CFN_PSEUDO_LIST
-        security_groups_ids = self.get_property(properties, 'SecurityGroupIds')
+        network_interface_ids = [ni.get('NetworkInterfaceId') for ni in network_interfaces]
+        network_interface_ids = network_interface_ids if check_iterable_has_value(network_interface_ids) else self.CFN_PSEUDO_LIST
+        security_group_ids_from_enis = flat_list([ni.get('GroupSet') for ni in network_interfaces])
+        security_group_ids_from_enis = set(sg for sg in security_group_ids_from_enis)
```
Comment (Tomerva) [inline cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py:23]: redundant list comprehension - you can just do `set(security_group_ids_from_enis)`, and you can merge this with previous line `set(flag_list .... ))`

File: cloudrail/knowledge/utils/utils.py [hunk 8]
```diff
@@ -137,6 +137,8 @@ def flat_list(list_of_lists: List[list]) -> list:
 def hash_list(data: list) -> int:
     return hash(str(data))
 
+def check_iterable_has_value(iterable: Iterable) -> bool:
+    return bool(iterable) and any(x is not None for x in iterable)
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/utils.py:141]: as previously discussed, the `any` part is sufficient 
image

File: cloudrail/knowledge/utils/utils.py [hunk 9]
```diff
@@ -137,6 +137,8 @@ def flat_list(list_of_lists: List[list]) -> list:
 def hash_list(data: list) -> int:
     return hash(str(data))
 
+def check_iterable_has_value(iterable: Iterable) -> bool:
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/utils.py:140]: rename to is_iterable_with_values

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py (hunk 5 above)
Comment (Tomerva) [reply cloudrail/knowledge/context/aws/resources_builders/cloudformation/ec2/cloudformation_ec2_builder.py]: @Imanuel-Mizrahi the above question still stands

**Review by idan-aviv — APPROVED** [review]

File: cloudrail/knowledge/utils/utils.py (hunk 8 above)
Comment (Imanuel-Mizrahi) [reply cloudrail/knowledge/utils/utils.py:141]: @Tomerva , but this is not good enough for a "set", I checked it, the tests failing.

**Review by Tomerva — APPROVED** [review]

**Pull request merged**